### PR TITLE
CMP: fix reporting error when no root CA cert update available

### PR DESCRIPTION
**Pull request by rajeev-0**
This pull request addresses the issue where an error occurs when the root CA certificate update is not available. In this case, the client should not classify it as an error.

##### Checklist

- [x] documentation is added or updated
- [x] tests are added or updated

### CODE REVIEW

File: crypto/cmp/cmp_asn.c [hunk 1]
```diff
@@ -287,23 +287,31 @@ OSSL_CMP_ITAV *OSSL_CMP_ITAV_new_rootCaKeyUpdate(const X509 *newWithNew,
...
-    if (newWithNew != NULL && (upd->newWithNew = X509_dup(newWithNew)) == NULL)
+
+    if ((upd->newWithNew = X509_dup(newWithNew)) == NULL)
         goto err;
     if (newWithOld != NULL && (upd->newWithOld = X509_dup(newWithOld)) == NULL)
         goto err;
     if (oldWithNew != NULL && (upd->oldWithNew = X509_dup(oldWithNew)) == NULL)
         goto err;
+
+ null_value:
```
Comment (DDvO) [inline crypto/cmp/cmp_asn.c:306]: Better not use `goto` for anything else but error handling.
So please use `if (newWithNew != NULL) { ... }`.

File: crypto/cmp/cmp_asn.c [hunk 2]
```diff
@@ -287,23 +287,31 @@ OSSL_CMP_ITAV *OSSL_CMP_ITAV_new_rootCaKeyUpdate(const X509 *newWithNew,
                                                  const X509 *oldWithNew)
 {
     OSSL_CMP_ITAV *itav;
-    OSSL_CMP_ROOTCAKEYUPDATE *upd = OSSL_CMP_ROOTCAKEYUPDATE_new();
+    OSSL_CMP_ROOTCAKEYUPDATE *upd = NULL;
 
+    if (newWithNew == NULL)
+        goto null_value;
+
+    upd = OSSL_CMP_ROOTCAKEYUPDATE_new();
```
Comment (DDvO) [inline crypto/cmp/cmp_asn.c]: `doc/man3/OSSL_CMP_ITAV_new_caCerts.pod` needs to be updated to cover the new use case for `OSSL_CMP_ITAV_get0_rootCaKeyUpdate()`.

File: crypto/cmp/cmp_asn.c (hunk 1 above)
Comment (rajeev-0) [reply crypto/cmp/cmp_asn.c]: done

File: test/recipes/80-test_cmp_http_data/test_commands.csv [hunk 3]
```diff
@@ -77,7 +77,7 @@ expected,description, -section,val, -cmd,val,val2, -cacertsout,val,val2, -infoty
 0,genm rootCaCert oldwithold empty file  , -section,, -cmd,genm,, BLANK,,, -infotype,rootCaCert,, -oldwithold, empty.txt     , -newwithnew, _RESULT_DIR/test.newwithnew.pem
 0,genm rootCaCert oldwithold random file , -section,, -cmd,genm,, BLANK,,, -infotype,rootCaCert,, -oldwithold, random.bin    , -newwithnew, _RESULT_DIR/test.newwithnew.pem
 0,genm rootCaCert oldwithold nonexistent , -section,, -cmd,genm,, BLANK,,, -infotype,rootCaCert,, -oldwithold, idontexist    , -newwithnew, _RESULT_DIR/test.newwithnew.pem
-0,genm rootCaCert oldwithold wrong       , -section,, -cmd,genm,, BLANK,,, -infotype,rootCaCert,, -oldwithold, signer.crt    , -newwithnew, _RESULT_DIR/test.newwithnew.pem
+1,genm rootCaCert oldwithold wrong       , -section,, -cmd,genm,, BLANK,,, -infotype,rootCaCert,, -oldwithold, signer.crt    , -newwithnew, _RESULT_DIR/test.newwithnew.pem
```
Comment (DDvO) [inline test/recipes/80-test_cmp_http_data/test_commands.csv:80]: oldwithold wrong -> oldwithold different

**Review by DDvO — CHANGES_REQUESTED** [review]
Just two more nits.

File: test/recipes/80-test_cmp_http_data/test_commands.csv (hunk 3 above)
Comment (rajeev-0) [reply test/recipes/80-test_cmp_http_data/test_commands.csv]: done

File: doc/man3/OSSL_CMP_ITAV_new_caCerts.pod [hunk 4]
```diff
@@ -59,7 +59,8 @@ If I<newWithOld> is not NULL, it assigns to I<*newWithOld> the internal pointer
 to the certificate contained in the newWithOld infoValue sub-field of I<itav>.
 If I<oldWithNew> is not NULL, it assigns to I<*oldWithNew> the internal pointer
 to the certificate contained in the oldWithNew infoValue sub-field of I<itav>.
-Each of these pointers will be NULL if the respective sub-field is not set.
+Each of these pointers will be set to NULL if update of root CA certificate is
+not included.
```
Comment (DDvO) [inline doc/man3/OSSL_CMP_ITAV_new_caCerts.pod:63]:
```suggestion
Each of these pointers will be set to NULL if no root CA certificate update 
is present or the respective sub-field is not included.
```

Comment (DDvO) [reply doc/man3/OSSL_CMP_ITAV_new_caCerts.pod:63]: Also the doc of `OSSL_CMP_ITAV_new_rootCaKeyUpdate()` needs to be extended:
An RootCaKeyUpdateContent structure is included only if *newWithNew* is not NULL.

This is what I actually meant in my yesterday's comment.

**Review by DDvO — CHANGES_REQUESTED** [review]
Two more little doc issues.

File: doc/man3/OSSL_CMP_ITAV_new_caCerts.pod (hunk 4 above)
Comment (rajeev-0) [reply doc/man3/OSSL_CMP_ITAV_new_caCerts.pod]: done

File: doc/man3/OSSL_CMP_ITAV_new_caCerts.pod [hunk 5]
```diff
@@ -49,6 +49,7 @@ the internal pointer to the certificate contained in the infoValue field.
 OSSL_CMP_ITAV_new_rootCaKeyUpdate() creates a new B<OSSL_CMP_ITAV> structure
 of type B<rootCaKeyUpdate> that includes an RootCaKeyUpdateContent structure
 with the optional I<newWithNew>, I<newWithOld>, and I<oldWithNew> certificates.
+An RootCaKeyUpdateContent structure is included only if newWithNew is not NULL.
```
Comment (DDvO) [inline doc/man3/OSSL_CMP_ITAV_new_caCerts.pod:52]: last nit: `newWithNew` -> `I<newWithNew>`

**Review by DDvO — APPROVED** [review]
Thanks for the fix

**Review by shahsb — APPROVED** [review]

**Comment by DDvO** [comment]
Merged to master, 3.3, and 3.2 - 
thanks @rajeev-0 for the contribution and @t8m for the review.

**Pull request closed without merging**